### PR TITLE
Function for reading GE Raw Data Server data

### PR DESCRIPTION
**Pull request by jfnielsen**
On branch main
	new file:   rds.jl

### CODE REVIEW

File: src/ge-mri/rds.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+"""
+    dat = loadrds(fid::IOStream, frames, frsize::Int, ncoil::Int;
+                  ptsize::Int = 2)
+
+Read multiple frames. Here, `frames` is an Int vector/array/range.
+
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil, length(frames)]`
+"""
+function loadrds(fid::IOStream, frames, frsize::Int, ncoil::Int;
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:54]:
```suggestion
function loadrds(fid::IOStream, frames::AbstractArray{<:Int}, frsize::Int, ncoil::Int;
```
This enforces the type requirement that you stated in words.

File: src/ge-mri/rds.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+"""
+function loadrds(fid::IOStream, frame::Int, frsize::Int, ncoil::Int; 
+    ptsize::Int = 2,
+)
+
+    if ptsize == 2
+        d = Array{Complex{Int16}}(undef, frsize*ncoil) # one readout
+    else
+        d = Array{Complex{Int32}}(undef, frsize*ncoil)
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:35]:
```suggestion
        d = Array{Complex{Int32}}(undef, frsize, ncoil)
```

File: src/ge-mri/rds.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+    seek(fid, 2*ptsize*(frame-1)*frsize*ncoil)
+
+    read!(fid, d)
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:40]:
```suggestion
    read!(fid, vec(d))
```
this uses a 1D "view" of d (like a pointer) so avoids duplicate allocations

File: src/ge-mri/rds.jl [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    reshape(d, frsize, ncoil)
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:42]:
```suggestion
    return d
```

Comment (JeffFessler) [reply src/ge-mri/rds.jl:42]: i haven't tested any of these suggestions btw, so some fixes may be needed :)

File: src/ge-mri/rds.jl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+)
+    if ptsize == 2
+        d = Array{Complex{Int16}}(undef, frsize, ncoil, length(frames)) 
+    else
+        d = Array{Complex{Int32}}(undef, frsize, ncoil, length(frames))
+    end
+
+    for i in 1:length(frames)
+        d[:, :, i] = loadrds(fid, frames[i], frsize, ncoil; ptsize = ptsize);
+    end
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:65]:
```suggestion
    d = stack([loadrds(fid, frame, frsize, ncoil; ptsize) for frame in frames]...)
```
i think this 1-liner will work!

File: src/ge-mri/rds.jl (hunk 4 above)
Comment (jfnielsen) [reply src/ge-mri/rds.jl:42]: It works :)

File: src/ge-mri/rds.jl (hunk 3 above)
Comment (jfnielsen) [reply src/ge-mri/rds.jl]: Neat, thanks!

File: src/ge-mri/rds.jl (hunk 5 above)
Comment (jfnielsen) [reply src/ge-mri/rds.jl:65]: It works without the '...'. 
```suggestion
d = stack([loadrds(fid, frame, frsize, ncoil; ptsize) for frame in frames])
```

**Comment by jfnielsen** [comment]
I believe it's ready to be merged now. Will add a test function at a later time.

File: src/ge-mri/rds.jl [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+Read multiple frames. Here, `frames` is an Int vector/array/range.
+
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil, length(frames)]`
+"""
+function loadrds(fid::IOStream, frames::AbstractArray{<:Int}, frsize::Int, ncoil::Int;
+    ptsize::Int = 2,
+)
+d = stack([loadrds(fid, frame, frsize, ncoil; ptsize) for frame in frames])
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:57]:
```suggestion
    d = stack([loadrds(fid, frame, frsize, ncoil; ptsize) for frame in frames])
```

File: src/ge-mri/rds.jl [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+"""
+function loadrds(fid::IOStream, frame::Int, frsize::Int, ncoil::Int; 
+    ptsize::Int = 2,
+)
+
+    if ptsize == 2
+        d = Array{Complex{Int16}}(undef, frsize, ncoil) # one readout
+    else
+        d = Array{Complex{Int32}}(undef, frsize, ncoil)
+    end
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:36]:
```suggestion
    T = ptsize == 2 ? Int16 : ptsize == 4 ? Int32 : throw(ArgumentError("ptsize $ptsize"))
    d = Array{Complex{T}}(undef, frsize, ncoil) # one readout
```

error checking and DRY principle

File: src/ge-mri/rds.jl [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+- `ptsize::Int`
+  Either 2 (real/complex value stored as Int16; default)
+  or 4 (Int32)
+  
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil]`
+
+"""
+function loadrds(fid::IOStream, frame::Int, frsize::Int, ncoil::Int; 
+    ptsize::Int = 2,
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:29]:
```suggestion
    T::Type{<:Union{Int16,Int32}} = Int16,
```

File: src/ge-mri/rds.jl [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    seek(fid, 2*ptsize*(frame-1)*frsize*ncoil)
+
+    read!(fid, vec(d))
+
+    return d
+end
+
+"""
+    dat = loadrds(fid::IOStream, frames, frsize::Int, ncoil::Int;
+                  ptsize::Int = 2)
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:44]:
```suggestion
                  T::Type{<:Union{Int16,Int32}} = Int16)
```

File: src/ge-mri/rds.jl [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+Read multiple frames. Here, `frames` is an Int vector/array/range.
+
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil, length(frames)]`
+"""
+function loadrds(fid::IOStream, frames::AbstractArray{<:Int}, frsize::Int, ncoil::Int;
+    ptsize::Int = 2,
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:52]:
```suggestion
    T::Type{<:Union{Int16,Int32}} = Int16,
```

File: src/ge-mri/rds.jl [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+)
+    d = stack([loadrds(fid, frame, frsize, ncoil; ptsize) for frame in frames])
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:54]:
```suggestion
    d = stack([loadrds(fid, frame, frsize, ncoil; sizeof(T)) for frame in frames])
```

File: src/ge-mri/rds.jl [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+- `ncoil::Int`
+  Number of receive coils
+
+option
+- `ptsize::Int`
+  Either 2 (real/complex value stored as Int16; default)
+  or 4 (Int32)
+  
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil]`
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:25]:
```suggestion
- `T::Type{<:Union{Int16,Int32}} = Int16` for the base type of (complex) data loaded
  
out
- `dat::Array{Complex{T}}` `[frsize, ncoil]`
```

File: src/ge-mri/rds.jl [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+"""
+function loadrds(fid::IOStream, frame::Int, frsize::Int, ncoil::Int; 
+    T::Type{<:Union{Int16,Int32}} = Int16,
+)
+
+    d = Array{Complex{T}}(undef, frsize, ncoil) # one readout
+
+    seek(fid, 2*ptsize*(frame-1)*frsize*ncoil)
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:34]:
```suggestion
    seek(fid, 2*sizeof(T)*(frame-1)*frsize*ncoil)
```

File: src/ge-mri/rds.jl [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+end
+
+"""
+    dat = loadrds(fid::IOStream, frames, frsize::Int, ncoil::Int;
+                  T::Type{<:Union{Int16,Int32}} = Int16)
+
+Read multiple frames. Here, `frames` is an Int vector/array/range.
+
+out
+- `dat::Array{Complex{Int16/Int32}}` `[frsize, ncoil, length(frames)]`
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:48]:
```suggestion
- `dat::Array{Complex{T}}` `[frsize, ncoil, length(frames)]`
```

File: src/ge-mri/rds.jl [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+"""
+function loadrds(fid::IOStream, frames::AbstractArray{<:Int}, frsize::Int, ncoil::Int;
+    T::Type{<:Union{Int16,Int32}} = Int16,
+)
+    d = stack([loadrds(fid, frame, frsize, ncoil; sizeof(T)) for frame in frames])
```
Comment (JeffFessler) [inline src/ge-mri/rds.jl:53]:
```suggestion
    d = stack([loadrds(fid, frame, frsize, ncoil; T) for frame in frames])
```

**Review by JeffFessler — APPROVED** [review]

**Comment by JeffFessler** [comment]
This needs a test for codecov and a version bump.  It will be easier to add a test after merging.

**Pull request merged**